tagmanager/src/tm_ctags_wrappers.h

```diff
@@ -0,0 +1,49 @@
+/*
+ *      tm_ctags_wrappers.h - this file is part of Geany, a fast and lightweight IDE
+ *
+ *      Copyright 2016 Jiri Techet <techet(at)gmail(dot)com>
+ *
+ *      This program is free software; you can redistribute it and/or modify
+ *      it under the terms of the GNU General Public License as published by
+ *      the Free Software Foundation; either version 2 of the License, or
+ *      (at your option) any later version.
+ *
+ *      This program is distributed in the hope that it will be useful,
+ *      but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *      GNU General Public License for more details.
+ *
+ *      You should have received a copy of the GNU General Public License along
+ *      with this program; if not, write to the Free Software Foundation, Inc.,
+ *      51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef TM_CTAGS_WRAPPERS
+#define TM_CTAGS_WRAPPERS
+
+#include <glib.h>
+
+#include "tm_parser.h"
+
+#include "entry.h" /* for sTagEntryInfo */
+
+
+G_BEGIN_DECLS
+
+typedef gboolean (*tm_ctags_callback) (const tagEntryInfo *const tag,
+	gboolean invalidate, void *user_data);
```

Comment:
> This is an odd type name for TM/Geany, we generally use CamelCase for those -- TMCtagsCallback or something?

OK.

> It also seems to have a strange signature only targeted at our very specific usage, although it suggest it's meant to be a generic API. By that I mean the invalidate parameter, that seem more like a step indicator that might be better served as an enum, or at least rather name it something like "initialize" or something more generic...

This isn't meant to be some final API and it's indeed targeted to Geany's specific uses. The thing about "invalidate" and multi-pass parsing is that I think it would be best to get rid of it completely and use just a single pass. From the API perspective it's really stupid because this means the API cannot be used as a stream - you cannot receive a tag, process it somehow and pass it somewhere (e.g. send over network) - you have to wait until the parsing is finished and only then you can be sure the tags are valid.

By the way, I tried the c.c parser on Geany+glib+glibc, then linux kernel and also the boost library sources and the second pass was never executed (plus all the Geany unit tests pass without it). I'm sure it can be useful in some very specific cases but I'm also sure one could introduce passes 2, 3, 4, 5... for other specific cases and this really isn't the way to go. So if there is some real ctags API in the future, IMO the way to go is to remove the "invalidate" parameter and multi-pass parsing.